### PR TITLE
feat(csi): extend csi snapshot to support backing image

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: [longhorn/longhorn#5005](https://github.com/longhorn/longhorn/issues/5005)

LEP: https://github.com/longhorn/longhorn/pull/5771

Summary:
- Users can use VolumeSnapshot with type `bi` to create/delete BackingImage
- Users can restore a Volume with a VolumeSnapshot (create a Volume with BackingImage)
- Users can restore a Volume with an existing BackingImage (pre-provision)
    - Create the VolumeSnapshot/VolumeSnapshotContent to associate to the existing BackingImage
- Users can restore a Volume with non-existing BackingImage (on-demand provision)
    - Create the VolumeSnapshot/VolumeSnapshotContent to provide BackingImage parameters
    - Create BackingImage based on the parameters when the users create PVC from the VolumeSnapshot

### CODE REVIEW

File: csi/controller_server.go [hunk 1]
```diff
@@ -578,6 +596,62 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 	return rsp, err
 }
 
+func (cs *ControllerServer) createCSISnapshotTypeLonghornBackingImage(req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	csiSnapshotName := req.GetName()
+	csiVolumeName := req.GetSourceVolumeId()
+	if len(csiVolumeName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume name must be provided")
```
Comment (innobead) [inline csi/controller_server.go:603]: If creating backing image via snapshot, why volume name is always required? three cases are here.

- create a backing image from a volume
- create a backing image from different sources (like download)
- create a backing image from a staticaly created snapshot content

Here, we only cover the first case? or please clarify this.

**Review by innobead — CHANGES_REQUESTED** [review]
Some comments and questions need to resolve.

File: csi/controller_server.go (hunk 1 above)
Comment (ChanYiLin) [reply csi/controller_server.go:603]: csi `createSnapsot` will only be triggered when `VolumeSnapshot` source pointing to a volume
```
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshot
metadata:
  name: test-snapshot-backing
spec:
  volumeSnapshotClassName: longhorn-snapshot-vsc
  source:
    persistentVolumeClaimName: test-vol
```
csi will then trigger the function to create snapshot with the given volume


On the other hand, if `VolumeSnapshot` source pointing to a `VolumeSnapshotContent`, csi won't trigger the `createSnapshot()` function
```
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshot
metadata:
  name: test-snapshot-on-demand-backing
spec:
  volumeSnapshotClassName: longhorn-snapshot-vsc
  source:
    volumeSnapshotContentName: test-on-demand-backing
```

We can only wait until user create PVC and using the `VolumeSnapshot` and the csi trigger `createVolume()` then we can provision the BackingImage

File: csi/controller_server.go [hunk 2]
```diff
@@ -282,20 +294,24 @@ func (cs *ControllerServer) checkAndPrepareBackingImage(volumeName, backingImage
...
+			return fmt.Errorf("volume %s backing image type %v is not supported via CSI", volumeName, bidsType)
+		case longhorn.BackingImageDataSourceTypeDownload:
+			if bidsParameters[longhorn.DataSourceTypeDownloadParameterURL] == "" {
+				return fmt.Errorf("volume %s missing parameters %v for preparing backing image",
+					volumeName, longhorn.DataSourceTypeDownloadParameterURL)
 			}
-			bidsType = string(longhorn.BackingImageDataSourceTypeDownload)
-			bidsParameters[longhorn.DataSourceTypeDownloadParameterURL] = volumeParameters["backingImageURL"]
+		case longhorn.BackingImageDataSourceTypeExportFromVolume:
+			if bidsParameters[longhorn.DataSourceTypeExportParameterExportType] == "" || bidsParameters[longhorn.DataSourceTypeExportParameterVolumeName] == "" {
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply csi/controller_server.go:309]: Yes, this enables user to use storageclass to provision backingimage with `export-from-volume` type

**Comment by ChanYiLin** [comment]
Since the modifications are related to original `snapshot`, `backup` volumesnapshot
let me run the e2e test to make sure they work fine

**Comment by ChanYiLin** [comment]
e2e test for csi snapshotter passed: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/3714

**Review by innobead — CHANGES_REQUESTED** [review]
Just a few minor comments needed to fix.

File: csi/controller_server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply csi/controller_server.go:310]: got it will update in this PR https://github.com/longhorn/website/pull/693

**Review by innobead — APPROVED** [review]
LGTM

File: csi/controller_server.go [hunk 3]
```diff
@@ -780,39 +881,75 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 	}
 
 	csiSnapshotType, sourceVolumeName, id := decodeSnapshotID(snapshotID)
-	if id == "" {
-		return nil, status.Errorf(codes.NotFound, "volume source snapshot %v is not found", snapshotID)
-	}
```
Comment (PhanLe1010) [inline csi/controller_server.go:785]: Could you please explain why we should delete this check?

Comment (ChanYiLin) [reply csi/controller_server.go:785]: the backing image has different snapshotID format
but we share the same decodeSnapshotID() function
- For BackingImage, the `sourceVolumeName`, `id` will be "", so we skip the check
- For Snapshot/Backup, we check in the below code https://github.com/longhorn/longhorn-manager/pull/1858/files#diff-4dbe40287acb148c5d0499c0adf977582db1e7b841f0d5d0943bd7e5e77b0cc8R890

**Pull request merged**